libs/shared/db/mysql/core/src/lib/core.ts

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { knex, Knex } from 'knex';
+import { promisify } from 'util';
+
+import { ConsoleLogger, Logger } from '@fxa/shared/log';
+import { localStatsD, StatsD } from '@fxa/shared/metrics/statsd';
+
+import { MySQLConfig } from './config';
+
+const REQUIRED_SQL_MODES = ['STRICT_ALL_TABLES', 'NO_ENGINE_SUBSTITUTION'];
+
+/**
+ * Setup a connection after knex establishes it, before its used.
+ *
+ * @param conn rawDriverConnection from knex
+ * @param done callback to indicate connection is ok
+ */
+async function setupConnection(conn: any, done: any) {
+  const query = promisify(conn.query).bind(conn);
+  try {
+    // Always communicate timestamps in UTC.
+    await query("SET time_zone = '+00:00'");
+    // Always use full 4-byte UTF-8 for communicating unicode.
+    await query('SET NAMES utf8mb4 COLLATE utf8mb4_bin;');
+    // Always have certain modes active. The complexity here is to
+    // preserve any extra modes active by default on the server.
+    // We also try to preserve the order of the existing mode flags,
+    // just in case the order has some obscure effect we don't know about.
+    const rows = await query('SELECT @@sql_mode AS mode');
+    const modes = rows[0].mode.split(',');
+    let needToSetMode = false;
+    for (const requiredMode of REQUIRED_SQL_MODES) {
+      if (modes.indexOf(requiredMode) === -1) {
+        modes.push(requiredMode);
+        needToSetMode = true;
+      }
+    }
+    if (needToSetMode) {
+      const mode = modes.join(',');
+      await query("SET SESSION sql_mode = '" + mode + "'");
+    }
+    done(null, conn);
+  } catch (err) {
+    done(err, conn);
+  }
+}
+
+function typeCasting(field: any, next: any) {
+  if (field.type === 'TINY' && field.length === 1) {
+    return field.string() === '1';
+  }
+  return next();
+}
+
+export function monitorKnexConnectionPool(pool: any, metrics: StatsD) {
+  metrics.increment('knex.pool_creation');
+  pool.on('acquireRequest', (eventId: any) => {
+    metrics.increment('knex.aquire_request');
+  });
+  pool.on('createRequest', (eventId: any) => {
+    metrics.increment('knex.create_request');
+  });
+  pool.on('destroyRequest', (eventId: any, resource: any) => {
+    metrics.increment('knex.destroy_request');
+  });
+  pool.on('destroyFail', (eventId: any, resource: any) => {
+    metrics.increment('knex.destroy_fail');
+  });
+}
+
+export function setupDatabase(
+  opts: MySQLConfig,
+  log: Logger = new ConsoleLogger(),
+  metrics: StatsD = localStatsD()
+): Knex {
+  const db = knex({
+    connection: {
+      typeCast: typeCasting,
+      charset: 'UTF8MB4_BIN',
```

Comment:
Just curious, why is it that this and line 26 are both needed?

Reply:
I believe its related to how the current connection handling, and stored procedure handling work. There's various parts of the process where the database may use different settings. We've had problems in the past where a stored procedure, or the connection, or a setting has a charset mismatch, which makes it not resolve some names consistently such that our result sets from queries miss records that should exist. As a result, we have perhaps gone overboard in setting this every possible place we can 'just to be sure'.

Reply:
Makes sense, better safe than sorry. 👍

Reply:
They are functionally equivalent.  Although I like `SET NAMES` more as it's more explicit.